### PR TITLE
[PY] feat: turn state

### PR DESCRIPTION
**Pull request by blackchoey**
## Linked issues

closes: #1079 

## Details

1. Create new TurnState related classes based on the JS implementation
2. Update the folder structure
3. Add unit tests for the new classes
4. Following behavior is different than JS due to the language difference:
    a. The `DefaultConversationState`, `DefaultUserState` and `DefaultTempState` are classes with a `dict` from `Storage`. Developers need to create child classes and add required properties to the class in order to access the data easily. Other solutions like using `dataclass` or `TypedDict` will introduce problems and bigger behavior changes, because the underlying data struct from `Storage` is a dict and `dataclass`/`TypedDict` cannot read/write that dict directly.
    b. Developer also need to inherit the `TurnState` class in order to return their own state classes.
5. Existing code that uses the old TurnState implementation is not updated, as it's a breaking change and updating existing code may contain a lot of throw away work. The old TurnState code needs to be cleaned up after we finish the breaking change. 

Remaining TODO items (not blocking):
- [x] Create follow up task to clean up old TurnState code #1164
- [x] Add docstring for the code

## Attestation Checklist

- [x] My code follows the style guidelines of this project

- I have checked for/fixed spelling, linting, and other errors
- I have commented my code for clarity
- I have made corresponding changes to the documentation (we use [TypeDoc](https://typedoc.org/) to document our code)
- My changes generate no new warnings
- I have added tests that validates my changes, and provides sufficient test coverage. I have tested with:
  - Local testing
  - E2E testing in Teams
- New and existing unit tests pass locally with my changes

### CODE REVIEW

File: python/packages/ai/tests/state/test_turn_state_entry.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+        self.entry.delete()
+        self.assertTrue(self.entry.is_deleted)
+
+    def test_is_deleted_when_access_again(self):
+        self.entry.delete()
+        self.assertTrue(self.entry.is_deleted)
+        self.assertEqual(self.entry.value, {})
+        self.assertFalse(self.entry.is_deleted)
+
+    def test_is_deleted_after_replace(self):
```
Comment (blackchoey) [inline python/packages/ai/tests/state/test_turn_state_entry.py:62]: This test case is failing due to bug https://github.com/microsoft/teams-ai/issues/1151. Need to confirm the expected behavior.

**Comment by blackchoey** [comment]
Also need to confirm behavior of below issue for TurnState: https://github.com/microsoft/teams-ai/issues/1153. Not blocking.

**Review by aacebo — APPROVED** [review]

**Comment by lilyydu** [comment]
@blackchoey @aacebo should we also add another ticket to update the existing samples from the breaking changes?

**Review by lilyydu — APPROVED** [review]

**Comment by blackchoey** [comment]
> @blackchoey @aacebo should we also add another ticket to update the existing samples from the breaking changes?

@lilyydu @aacebo created https://github.com/microsoft/teams-ai/issues/1175 for this. We could triage the new items offline or in our next scrum meeting.

**Pull request merged**